src/diem/offchain/funds_pull_pre_approval_command_state.py

```diff
@@ -0,0 +1,41 @@
+# Copyright (c) The Diem Core Contributors
+# SPDX-License-Identifier: Apache-2.0
+
+import typing
+
+from .types import FundPullPreApprovalStatus
+
+Status = str
+
+_transitions: typing.Dict[Status, typing.List[Status]] = {
+    FundPullPreApprovalStatus.pending: [
+        FundPullPreApprovalStatus.pending,
+        FundPullPreApprovalStatus.valid,
+        FundPullPreApprovalStatus.rejected,
+        FundPullPreApprovalStatus.closed,
+    ],
+    FundPullPreApprovalStatus.valid: [
+        FundPullPreApprovalStatus.valid,
+        FundPullPreApprovalStatus.closed,
+    ],
+    FundPullPreApprovalStatus.rejected: [
+        FundPullPreApprovalStatus.rejected,
+    ],
+    FundPullPreApprovalStatus.closed: [
+        FundPullPreApprovalStatus.closed,
+    ],
```

Comment:
Why every state can go to itself? Allow partial update? At least close should not be allowed. 
I suggest we start with no partial update which is direction we go for dip-1.

Reply:
This behavior stems from the protocol idempotency requirement. It has been discussed in several meetings and specifically [stated in DIP-1](https://github.com/diem/dip/blob/master/dips/dip-1.mdx#network-error-handling):

> Upon receipt of a Command that has already been processed (resulting in a success response or a Command error), the receiving side must reply with the same response as was previously issued (successful commands, or commands that fail with a command error, are idempotent).

Reply:
for the case reply with same response, it should not be part of the state machine, it can be simply checked by if old == new.